ehrapy/tools/cohort_tracking/_cohort_tracker.py

```diff
@@ -0,0 +1,332 @@
+import copy
+from typing import Any, Union
+
+import graphviz
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import seaborn as sns
+from scanpy import AnnData
+from tableone import TableOne
+
+
+def _check_columns_exist(df, columns):
+    if not all(col in df.columns for col in columns):
+        missing_columns = [col for col in columns if col not in df.columns]
+        raise ValueError(f"Columns {missing_columns} not found in dataframe.")
+
+
+# from tableone: https://github.com/tompollard/tableone/blob/bfd6fbaa4ed3e9f59e1a75191c6296a2a80ccc64/tableone/tableone.py#L555
+def _detect_categorical_columns(data) -> list:
+    # assume all non-numerical and date columns are categorical
+    numeric_cols = set(data._get_numeric_data().columns.values)
+    date_cols = set(data.select_dtypes(include=[np.datetime64]).columns)
+    likely_cat = set(data.columns) - numeric_cols
+    # mypy absolutely looses it if likely_cat is overwritten to be a list
+    likely_cat_no_dates = list(likely_cat - date_cols)
+
+    # check proportion of unique values if numerical
+    for var in data._get_numeric_data().columns:
+        likely_flag = 1.0 * data[var].nunique() / data[var].count() < 0.005
+        if likely_flag:
+            likely_cat_no_dates.append(var)
+    return likely_cat_no_dates
+
+
+class CohortTracker:
+    def __init__(self, adata: AnnData | pd.DataFrame, columns: list = None, categorical: list = None, *args: Any):
+        """Track cohort changes over multiple filtering or processing steps.
+
+        This class offers functionality to track and plot cohort changes over multiple filtering or processing steps,
+        enabling the user to monitor the impact of each step on the cohort.
+
+        Tightly interacting with the `tableone` package [1].
+        Args:
+            adata: :class:`~anndata.AnnData` or :class:`~pandas.DataFrame` object to track.
```

Comment:
I'm not sure whether we need DataFrame support but let's keep it for now

Reply:
Main motivation was that this tracker might be of interest for people also not working with our format, but plain DataFrame. Lower usage barrier this way

Reply:
I actually think that we WANT to force people into our format. The less things we're supporting, the easier it is for us to perform maintenance and the less potential for bugs. Have you noticed that most Google products or most modern apps in general have almost no settings? That's for this reason. Don't give people options unless there's really good justification

Reply:
By virtue of the last comment, removed pd.DataFrame support - can always add later or upon request, not a big thing one or the other way